contributingGuides/TYPESCRIPT_STYLE.md

```diff
@@ -0,0 +1,897 @@
+type Colors = TupleToUnion<typeof COLORS>;
+```
+
+Defining union types using an object offers several benefits. This approach it's a great choice when **enum-like identifiers are needed**, while maintaining one source of truth and improving readability.
+
+```ts
+// Bad
+const COLORS = {
+    Red: 'red',
+    Green: 'green',
+    Blue: 'blue',
+} as const;
+type Colors = 'red' | 'green' | 'blue';
+
+// Good
+import { ValueOf } from 'type-fest';
+
+const COLORS = {
+    Red: 'red',
+    Green: 'green',
+    Blue: 'blue',
+} as const;
+type Colors = ValueOf<typeof COLORS>
+```
+
+If using enum is necessary, utilize string enum. Do not use numeric and heterogenous enums as they auto increment its numeric values in unexpected ways, are hard to understand when they have mixed value types and have reverse mappings. This can help prevent potential bugs at runtime.
+
+```ts
+// Bad
+enum Colors1 {
+  Red,   // when not assigning anything, the value will start with 0.
+  Green, // and will increment, 1.
+  Blue,  // 2.
+}
+
+enum Colors2 {
+  Red = 10,            // 10.
+  Green,               // will increment to 11.
+  Blue = 30,           // 30.
+  Yellow,              // same, will increment to 31.
+  Magenta = 'magenta', // "magenta".
+  Brown = 100,         // 100.
+  Black,               // 101.
+}
+
+// Bad - We also end up having reverse mappings when using numeric values in enums.
+Colors2.Green; // 11.
+Colors2[11];   // "Green".
+
+// Good
+enum Colors {
+  Red = 'red',
+  Green = 'green',
+  Blue = 'blue',
+}
+```
+
+## Usage of type `any` and `unknown`
+
+Usage of `any` is forbidden unless there is a very good reason for it. The `any` type bypasses the TypeScript type-checking system, thus making the code less safe and more prone to bugs. Any usage of `any` type must be followed with a comment explaining why you are using it.
+
+```ts
+// Bad
+function processAPIResponse(response: any) {
+    // your code
+}
+
+// Good
+// As we are just going to persist the response in Onyx and
+// it would be very hard to type all possible data structures,
+// we just leave `response` as `any`.
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+function processAPIResponse (response: any) {
+    // your code
+}
+```
+
+The `any` type allows assignment to all types and dereference of any property, which is undesirable and should be avoided. Instead, and in most cases, use the `unknown` type which expresses a similar concept and is much safer as it requires narrowing the type before using it. 
+
+When you know that the type structure is a object but you don't have context about the content, use `Record<string, unknown>`. Note that numeric keys are allowed with the `Record<string, unknown>` type. Numeric keys are implicitly converted to strings during property assignment and access. To read more about this, see [TypeScript documentation](https://www.typescriptlang.org/docs/handbook/2/objects.html#dynamically-adding-properties).
```

Comment:
```suggestion
When you know that the type structure is an object, but you don't have context about the content, use `Record<string, unknown>`. Note that numeric keys are allowed with the `Record<string, unknown>` type. Numeric keys are implicitly converted to strings during property assignment and access. To read more about this, see [TypeScript documentation](https://www.typescriptlang.org/docs/handbook/2/objects.html#dynamically-adding-properties).
```

Reply:
I also think it might be good to include an example showing `Record<string, unknown>`, because that's not included in the example that follows.

Reply:
Changed the text and added an example.